core/shared/src/main/scala/zio/Ref.scala

```diff
@@ -18,126 +18,64 @@ package zio
 
 import java.util.concurrent.atomic.AtomicReference
 
-/**
- * A mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
- * a volatile `var`, augmented with atomic operations, which make it useful as a
- * reasonably efficient (if low-level) concurrency primitive.
- *
- * {{{
- * for {
- *   ref <- Ref.make(2)
- *   v   <- ref.update(_ + 3)
- *   _   <- console.putStrLn("Value = " + v) // Value = 5
- * } yield ()
- * }}}
- */
-final class Ref[A] private (private val value: AtomicReference[A]) extends AnyVal with Serializable {
-
-  /**
-   * Reads the value from the `Ref`.
-   */
-  final def get: UIO[A] = IO.effectTotal(value.get)
+object Ref extends Serializable {
 
   /**
-   * Atomically modifies the `Ref` with the specified function, which computes
-   * a return value for the modification. This is a more powerful version of
-   * `update`.
+   * Creates a new `Ref` with the specified value.
    */
-  final def modify[B](f: A => (B, A)): UIO[B] = IO.effectTotal {
-    var loop = true
-    var b: B = null.asInstanceOf[B]
-
-    while (loop) {
-      val current = value.get
-
-      val tuple = f(current)
-
-      b = tuple._1
-
-      loop = !value.compareAndSet(current, tuple._2)
-    }
-
-    b
+  final def make[A](a: A): UIO[Ref[A]] = IO.effectTotal {
+    unsafeCreate(a)
   }
 
-  /**
-   * Atomically modifies the `Ref` with the specified partial function, which computes
-   * a return value for the modification if the function is defined in the current value
-   * otherwise it returns a default value.
-   * This is a more powerful version of `updateSome`.
-   */
-  final def modifySome[B](default: B)(pf: PartialFunction[A, (B, A)]): UIO[B] = IO.effectTotal {
-    var loop = true
-    var b: B = null.asInstanceOf[B]
+  private final def unsafeCreate[A](a: A): Ref[A] =
+    new Ref[A] {
+      val value: AtomicReference[A] = new AtomicReference[A](a)
 
-    while (loop) {
-      val current = value.get
+      /**
+       * See [[ZRef#get]]
+       */
+      final def get: UIO[A] = IO.effectTotal(value.get)
 
-      val tuple = pf.applyOrElse(current, (_: A) => (default, current))
+      /**
+       * See [[ZRef#set]]
+       */
+      final def set(a: A): UIO[Unit] = IO.effectTotal(value.set(a))
 
-      b = tuple._1
-
-      loop = !value.compareAndSet(current, tuple._2)
-    }
-
-    b
-  }
+      /**
+       * See [[ZRef#setAsync]]
+       */
+      final def setAsync(a: A): UIO[Unit] = IO.effectTotal(value.lazySet(a))
 
-  /**
-   * Writes a new value to the `Ref`, with a guarantee of immediate
-   * consistency (at some cost to performance).
-   */
-  final def set(a: A): UIO[Unit] = IO.effectTotal(value.set(a))
+      private[zio] def unsafeUpdate[E](f: Any => Any): UIO[A] = IO.effectTotal {
```

Comment:
spurious `E`?

Reply:
should I return an `IO[E,A]` ?  Ref return `UIO` by definition, not sure what is the convention in this case

Reply:
I only saw the `unsafeUpdate` definition after. Apparently the scala compiler allows you to change the return type, I didn't know that 🤔 I'm inclined to say it doesn't really matter then.

Reply:
I'd delete the `E` since it isn't used, and `UIO[A]` is a subtype of `IO[E, A]` for all `E`.